presto-hive/src/main/java/com/facebook/presto/hive/HiveTableLayoutHandle.java

```diff
@@ -147,7 +147,21 @@ protected HiveTableLayoutHandle(
         this.layoutString = requireNonNull(layoutString, "layoutString is null");
         this.requestedColumns = requireNonNull(requestedColumns, "requestedColumns is null");
         this.partialAggregationsPushedDown = partialAggregationsPushedDown;
-        this.appendRowNumberEnabled = appendRowNumberEnabled;
+        // if $row_id is requested, we need the row numbers whether they're explicitly requested or not
+        // TODO is it possible someone needs the row_numbers and row IDs, and thus we need a block for each?
+        // in which case we might need to know whether the row numbers were explicitly requested, and thus
+        // need a separate variable for appending row IDs
+        // TODO verify that $row_id can appear in any of these column lists
+        if (dataColumns.stream().anyMatch(column -> column.getName().equals("$row_id"))
+                || predicateColumns.values().stream().anyMatch(column -> column.getName().equals("$row_id"))) {
```

Comment:
Do we need to do something to make sure that the $row_id is not used as a pushdown filter for selective readers?

Reply:
fixed by storing separate flags for appendRowNumberEnabled and appendRowId. If either is true, we ask the ORC reader for row IDs.